### PR TITLE
Flatten `planner` tables and refactor making edits to the roadmap

### PR DESCRIPTION
**Pull request by Awesome-E**
## Description

**Backend Changes:**
- Changed it so that instead of storing `years` directly as a (json) column for `planner`, there is now a `planner_year`, `planner_quarter`, and `planner_course` table for each of their respective data.
   - While this does take more steps to write to the database, it allows us to perform such updates incrementally – this enables us to implement things in the future such as auto-save and more advanced queries on courses in planners.
   - the `years` column must be deleted in a future database migration/pr since we need to merge the current changes and migrate data before erasing anything.
- Migration script to copy data from the `years` column of `planner` into each of their own tables.
- Changed the `trpc.roadmaps.save` route to take in a list of incremental changes and write the appropriate rows to the table. This makes it so we're not updating every planner every time, but instead only the courses, quarters, years, or planners affected.

**Frontend Changes:**
- Implement a system where all edits to the roadmap now create a "set of changes" rather than writing directly to the roadmap. Each change is then added to a "change list // revision stack", and when added, it gets "applied" to the roadmap.
   - By storing things in a "revision stack", we enable things like roadmap versioning (allowing undo/redo in the future) and allow every change to be unit tested
- Added a system to compare two roadmaps and create an incremental list of changes (new items [quarters/years/planners], updated items, and deleted items).
- When dealing with planners that haven't been saved to the db yet, we now create temporary (negative) IDs so we can associate data with them (i.e. associate a year that's created in a new planner) in our "sets of incremental changes".
   - The temporary IDs are sent to the backend when saving to account due to the need to associate data as described above, but they get replaced with permanent, positive IDs.
   - The temporary IDs are stored in localStorage when saving locally.

**Codebase/Stack Changes:**
- Wrote initial setup for Jest unit tests, and with that, added unit tests for applying revisions to a roadmap.
- Created "test skeletons" for all other helper functions in this PR, such as creating the revision data from something like drag-n-drop events. These tests are not implemented yet, but we have outlined what checks should pass for each one.
- Added a GitHub Action to run the test suite

## Screenshots

N/A

## Test Plan

- Unit tests should pass (hopefully)
- Any changes to the roadmap should continue to work on the frontend and to save properly (both localStorage and account).
  - [ ] Adding courses to a quarter
  - [ ] Removing courses from a quarter
  - [ ] Reordering courses within a quarter
  - [ ] Adding a quarter to a year that already exists
  - [ ] Removing a quarter from a year
  - [ ] Adding a new year to a planner (this also creates new quarters), with courses.
  - [ ] Removing a year from a planner (this also deletes quarters)
  - [ ] Renaming a year in a planner
  - [ ] Changing the start year in a planner
  - [ ] Creating a planner
  - [ ] Deleting a planner
  - [ ] Renaming a planner

Each of these are the "smallest individual edit" that gets recorded in each revision to the user's roadmap, and PeterPortal should be able to handle all of these changes and save them properly. Ideally, this is tested both while saving in between some actions and while applying all actions before saving.

Some other edge cases to test:
- [ ] When the user makes changes while logged out, then logs back in, there is a prompt to overwrite the account roadmap (by clicking "This Device"). Clicking this causes a different type of save to happen (it's an overwrite which deletes all existing planners from that user first), and this should still work as expected.

## Issues

N/A

### CODE REVIEW

File: site/src/types/roadmap.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  type: 'quarter';
+  plannerId: number;
+  startYear: number;
+  before: PlannerQuarterChangeData;
+  after: PlannerQuarterChangeData;
+}
+
+export type RoadmapEdit = PlannerEdit | PlannerYearEdit | PlannerQuarterEdit;
+
+// Revisions have multiple changes
```
[the remark this answers is not in the document]
Comment (Awesome-E) [reply site/src/types/roadmap.ts:44]: It is true that revisions have multiple edits, though I suppose the typedef below the comment speaks for itself. Unless you're referring to the RoadmapEdit type... if that's the case I'm not sure what you're referring to.
